### ISSUE TITLE
database stuck in a restart loop - docker-entrypoint.sh: Permission denied

### ISSUE THREAD

**Issue by paulverket** [adriankumpf/teslamate#756]
**Describe the bug**

I've been running teslamate successfully for a while until today when I had to reboot the machine. The database was stuck in a restart loop. 

```
verket@pi3:~/Projects/teslaMateLocal$ sudo docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                            PORTS               NAMES
92928027e018        postgres:12         "docker-entrypoint.s…"   2 minutes ago       Restarting (126) 35 seconds ago                       teslamatelocal_db_1

```
The logs showed the problem:
```
verket@pi3:~/Projects/teslaMateLocal$ sudo docker logs --tail 50 --follow --timestamps 92928027e018
2020-05-26T00:52:27.740455484Z bash: /usr/local/bin/docker-entrypoint.sh: Permission denied
2020-05-26T00:52:34.780288596Z bash: /usr/local/bin/docker-entrypoint.sh: Permission denied

```
If I weren't such a Docker newbie I'd know how to fix this -- sorry.

I have a few db backups so I don't mind blowing things away and starting again. I tried to do that with a combination of:
```
sudo docker-compose  down
sudo docker-compose rm db
sudo docker volume rm ...
```

but still get the same problem when I do `sudo docker-compose up -d db`

**Environment**

- TeslaMate version: 1.19.3
- Type of installation: Docker 18.09.9
- OS TeslaMate is installed on: Ubuntu 19.10 (64 bit) on Raspberry pi 3
- User OS & Browser: Mac OS 10.15.4, Safari 13.1

Thanks in advance...
Paul

**Comment by paulverket** [comment]
Continued trying to fix the db looping issue by deleting everything:
```
sudo docker system prune -a
sudo docker volume prune
```

Downloading and running just the db continued with the restart loop.
`sudo docker-compose up -d db`

Docker 18.09.9 was installed as a snap. I'm debating whether I should remove it and install with apt-get.

**Comment by Dulanic** [comment]
If I had to guess it has something to do with your file permissions wherever your mounted volume location is. Can you post your docker-compose.yml? This question may need to move more to the postgresql docker area.

This is the script it's trying to run from within the container. https://github.com/docker-library/postgres/blob/master/12/docker-entrypoint.sh

**Comment by paulverket** [comment]
Tried removing and reinstalling docker from a snap: 
```
sudo snap remove docker
sudo snap install docker
```
Reran `sudo docker-compose up -d db` but still get restarts:
```
$ sudo docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED              STATUS                            PORTS               NAMES
4cfa46569cee        postgres:12         "docker-entrypoint.s…"   About a minute ago   Restarting (126) 18 seconds ago                       teslamatelocal_db_1
$ sudo docker logs --tail 50 --follow --timestamps 4cfa46569cee
2020-05-27T16:51:11.210201098Z bash: /usr/local/bin/docker-entrypoint.sh: Permission denied
2020-05-27T16:51:21.173613816Z bash: /usr/local/bin/docker-entrypoint.sh: Permission denied
```

Here's the docker-compose.yml file:
```
version: '3'

services:
  teslamate:
    image: teslamate/teslamate:latest
    restart: always
    environment:
      - DATABASE_USER=teslamate
      - DATABASE_PASS=xxx
      - DATABASE_NAME=teslamate
      - DATABASE_HOST=db
      - MQTT_HOST=mosquitto
    ports:
      - 4000:4000
    volumes:
      - ./import:/opt/app/import
    cap_drop:
      - all

  db:
    image: postgres:12
    restart: always
    environment:
      - POSTGRES_USER=teslamate
      - POSTGRES_PASSWORD=xxx
      - POSTGRES_DB=teslamate
    volumes:
      - teslamate-db:/var/lib/postgresql/data

  grafana:
    image: teslamate/grafana:latest
    restart: always
    environment:
      - DATABASE_USER=teslamate
      - DATABASE_PASS=xxx
      - DATABASE_NAME=teslamate
      - DATABASE_HOST=db
    ports:
      - 3000:3000
    volumes:
      - teslamate-grafana-data:/var/lib/grafana

  mosquitto:
    image: eclipse-mosquitto:1.6
    restart: always
    ports:
      - 1883:1883
    volumes:
      - mosquitto-conf:/mosquitto/config
      - mosquitto-data:/mosquitto/data

volumes:
    teslamate-db:
    teslamate-grafana-data:
    mosquitto-conf:
    mosquitto-data:
```

Thanks!

**Comment by tacotran** [comment]
Can you try some manual steps?
```
#Ensure container is stopped
docker-compose stop database
docker-compose rm database

#run database manually
docker run -it --entrypoint /bin/bash postgres:12
#inside container now
ls -al /
ls -al usr/local/bin/docker-entrypoint.sh
sh usr/local/bin/docker-entrypoint.sh
```

**Comment by paulverket** [comment]
```
verket@pi3:~/Projects/teslaMateLocal$ sudo docker-compose stop db
verket@pi3:~/Projects/teslaMateLocal$ sudo docker-compose rm db
Going to remove teslamatelocal_db_1
Are you sure? [yN] y
Removing teslamatelocal_db_1 ... done
verket@pi3:~/Projects/teslaMateLocal$ sudo docker run -it --entrypoint /bin/bash postgres:12
bash: /root/.bashrc: Permission denied
root@b4136ea68a99:/# ls -al /
total 72
drwxr-xr-x   1 root root 4096 May 28 02:52 .
drwxr-xr-x   1 root root 4096 May 28 02:52 ..
drwxr-xr-x   2 root root 4096 May 14 14:50 bin
drwxr-xr-x   2 root root 4096 May  2 16:39 boot
drwxr-xr-x   5 root root  360 May 28 02:52 dev
drwxr-xr-x   2 root root 4096 May 16 06:28 docker-entrypoint-initdb.d
lrwxrwxrwx   1 root root   34 May 16 06:29 docker-entrypoint.sh -> usr/local/bin/docker-entrypoint.sh
-rwxr-xr-x   1 root root    0 May 28 02:52 .dockerenv
drwxr-xr-x   1 root root 4096 May 28 02:52 etc
drwxr-xr-x   2 root root 4096 May  2 16:39 home
drwxr-xr-x   1 root root 4096 May 16 06:27 lib
drwxr-xr-x   2 root root 4096 May 14 14:50 media
drwxr-xr-x   2 root root 4096 May 14 14:50 mnt
drwxr-xr-x   2 root root 4096 May 14 14:50 opt
dr-xr-xr-x 147 root root    0 May 28 02:52 proc
drwx------   1 root root 4096 May 16 06:27 root
drwxr-xr-x   1 root root 4096 May 16 06:28 run
drwxr-xr-x   2 root root 4096 May 14 14:50 sbin
drwxr-xr-x   2 root root 4096 May 14 14:50 srv
dr-xr-xr-x  12 root root    0 May 28 02:52 sys
drwxrwxrwt   1 root root 4096 May 16 06:28 tmp
drwxr-xr-x   1 root root 4096 May 14 14:50 usr
drwxr-xr-x   1 root root 4096 May 14 14:50 var
root@b4136ea68a99:/# ls -al /usr/local/bin/docker-entrypoint.sh
-rwxrwxr-x 1 root root 10927 May 16 06:27 /usr/local/bin/docker-entrypoint.sh
root@b4136ea68a99:/# sh /usr/local/bin/docker-entrypoint.sh
sh: 0: Can't open /usr/local/bin/docker-entrypoint.sh
root@b4136ea68a99:/# ls -ld /usr/local
drwxr-xr-x 1 root root 4096 May 14 14:50 /usr/local
root@b4136ea68a99:/# ls -ld /usr/local/bin
drwxr-xr-x 1 root root 4096 May 16 06:29 /usr/local/bin
root@b4136ea68a99:/# wc -l /usr/local/bin/docker-entrypoint.sh
wc: /usr/local/bin/docker-entrypoint.sh: Permission denied
```

Thanks for the help and education!
Paul

**Comment by paulverket** [comment]
I didn't think of SELinux or AppArmor -- thanks!

In terminal 1:
```
verket@pi3:~$ sudo docker run -it --entrypoint /bin/bash postgres:12
[sudo] password for verket: 
root@0f6e848e0659:/# ls -al usr/local/bin/docker-entrypoint.sh
-rwxrwxr-x 1 root root 10927 May 16 06:27 usr/local/bin/docker-entrypoint.sh
root@0f6e848e0659:/# sh usr/local/bin/docker-entrypoint.sh
usr/local/bin/docker-entrypoint.sh: 2: set: Illegal option -o pipefail
```

In terminal 2:
```
verket@pi3:~$ sudo sestatus
[sudo] password for verket: 
SELinux status:                 disabled
verket@pi3:~$ sudo aa-status
apparmor module is loaded.
39 profiles are loaded.
39 profiles are in enforce mode.
   /sbin/dhclient
   /snap/core/9068/usr/lib/snapd/snap-confine
   /snap/core/9068/usr/lib/snapd/snap-confine//mount-namespace-capture-helper
   /usr/bin/man
   /usr/lib/NetworkManager/nm-dhcp-client.action
   /usr/lib/NetworkManager/nm-dhcp-helper
   /usr/lib/connman/scripts/dhclient-script
   /usr/lib/snapd/snap-confine
   /usr/lib/snapd/snap-confine//mount-namespace-capture-helper
   /usr/sbin/tcpdump
   docker-default
   lsb_release
   man_filter
   man_groff
   nvidia_modprobe
   nvidia_modprobe//kmod
   snap-update-ns.core
   snap-update-ns.docker
   snap-update-ns.lxd
   snap.core.hook.configure
   snap.docker.compose
   snap.docker.docker
   snap.docker.dockerd
   snap.docker.help
   snap.docker.hook.install
   snap.docker.hook.post-refresh
   snap.docker.machine
   snap.lxd.activate
   snap.lxd.benchmark
   snap.lxd.buginfo
   snap.lxd.check-kernel
   snap.lxd.daemon
   snap.lxd.hook.configure
   snap.lxd.hook.install
   snap.lxd.hook.remove
   snap.lxd.lxc
   snap.lxd.lxc-to-lxd
   snap.lxd.lxd
   snap.lxd.migrate
0 profiles are in complain mode.
5 processes have profiles defined.
5 processes are in enforce mode.
   /usr/bin/bash (1892) docker-default
   /snap/docker/427/bin/docker (1828) snap.docker.docker
   /snap/docker/427/bin/dockerd (1220) snap.docker.dockerd
   /snap/docker/427/bin/containerd (1464) snap.docker.dockerd
   /snap/docker/427/bin/containerd-shim (1875) snap.docker.dockerd
0 processes are in complain mode.
0 processes are unconfined but have a profile defined.
```

On a side note, while checking on using apt instead of snap packages, I ran across this disclaimer: _Using docker.io on non-amd64 hosts is not supported at this time. Please be careful when using it on anything besides amd64._ I'm using arm 64 (pi 3)

**Comment by paulverket** [comment]
Removed the docker snap from pi3 and installed docker from docker.com with https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository. 

It doesn’t include docker-compose. The docker web page pointer to GitHub’s docker-compose doesn’t work because they don’t have a aarch64 build. Used apt-get install to get docker-compose from ubuntu's repo after verifying that it doesn’t also install the docker engine (only recommended). 

Tried the interactive test with: `sudo docker run -it --entrypoint /bin/bash postgres:12` but it still failed on: 
```
sh usr/local/bin/docker-entrypoint.sh
usr/local/bin/docker-entrypoint.sh: 2: set: Illegal option -o pipefail
```
Checking the .sh file showed that it required bash. Running it with bash worked fine. Brought up the db and it stays up!

Restored the DB and loaded the rest of the yaml file. **I'm back up!** Thanks @tacotran and @Dulanic for your help!

Postscript: For the SELinux/AppArmor test I had removed the docker snap and reinstalled before the test. Note that I got the same `usr/local/bin/docker-entrypoint.sh: 2: set: Illegal option -o pipefail` error so I might have been able to recover from that point.
